### DISCUSSION TITLE
Update libjpeg-turbo to universal binary

### DISCUSSION THREAD

**Discussion by AndySomogyi** [Homebrew/discussions#1225] [Everyday usage]
I'm trying to distribute some binaries, and I always statically link to Brew packages. 

But now, I need to start shipping universal binaries (x86-64 and arm), and looks like none of the Brew packages are universal binary, so my code fails on ARM. 

Could we update the Brew packages to universal binary? or at least have some sort of "--universal" option?

**Comment by carlocab** [comment]
We specifically avoid trying to build universal binaries because of bad experiences with them in the past.

See, for example, the following discussion (and the issues linked there) for details: https://github.com/Homebrew/discussions/discussions/412

**Reply by AndySomogyi** [reply]
So would you suggest that we stop using Brew and switch everything to git submodules and create individual build scripts for all our third party libraries?

**Reply by carlocab** [reply]
If that's the simplest alternative for you, then sure. You could also try stitching together the binary artefacts in the ARM and Intel Big Sur bottles with `lipo`, but I'm not certain this will work exactly how you'd want. Switching to git submodules + individual build scripts might be more reliable.

**Reply by AndySomogyi** [reply]
Wow. 

So you're saying Brew is dead? Is that right?

Because without universal binary support, Brew is simply useless. Apple, not us developers decided to switch to ARM, thus we developers need to distribute universal binaries. 

If I'm reading this correctly, you're saying that Brew is going to ignore the fact that we will have a mix of ARM and x86 Macs for a very long time to come.

**Reply by carlocab** [reply]
If by "dead" you mean "has no intention of supporting universal binaries", then yes. Though this choice of language just reminds me of something Mark Twain once said.

Perhaps this makes `brew` useless _for you_, but that is a far cry from being useless, full-stop.

>If I'm reading this correctly, you're saying that Brew is going to ignore the fact that we will have a mix of ARM and x86 Macs for a very long time to come.

This is incorrect.

**Reply by alebcay** [reply]
> Because without universal binary support, Brew is simply useless.

For who? I'd like to hope that [some of these 13 million install events in the past 30 days](https://formulae.brew.sh/analytics/os-version/30d/) are useful to _somebody_ because otherwise that's a lot of wasted bandwidth and electricity.

> If I'm reading this correctly, you're saying that Brew is going to ignore the fact that we will have a mix of ARM and x86 Macs for a very long time to come.

Well, I would like to posit that you're _not_ reading this correctly. Homebrew is expecting there to be a mix of ARM and x86 Macs for a long time moving forward. ARM support was not just monkey-patched into Homebrew for the sake of supporting Big Sur. The decision to ship separate binaries was one that was deliberated over internally for months leading up to the release - those decisions are elaborated in the linked discussion (and from links in that discussion).

Keep in mind that there is no _technical_ limitation in Homebrew that's stopping you from building universal binaries. You just have to maintain your own formula, in which the universal binary version of the software is built (and you can then statically link to that). The point is that Homebrew doesn't want to support this configuration as our default (for the reasons mentioned in that linked discussion); the onus of using Homebrew to accomplish your particular use case is solely yours.

> So you're saying Brew is dead?

I wish you'd told us before last November! It would've saved a lot of maintainer effort over the past six months.

**Reply by MikeMcQuaid** [reply]
> Because without universal binary support, Brew is simply useless. 

No.

**Comment by danielnachun** [comment]
> I'm trying to distribute some binaries, and I always statically link to Brew packages.

This is not and likely never will be a supported use of Homebrew (other maintainers please feel free to correct me).  If you want to use Homebrew to distribute binaries to your end users, you should submit a formula so that we can build it in CI, and let us handle distributing the right binary to each user based their system architecture.  This is how Linux package managers work, and that is generally the model we follow.

If you can't submit a formula because your code is proprietary, then you'll need to build the libraries as universal binaries from source.  You could use `brew tap` and `brew extract` on the libraries and modify the formulae to build as universal binaries, or you could build them from source outside Homebrew.  You can then package your binaries in an installer and create a Cask which end users can install with Homebrew.  In that case we don't distribute the binaries for you, we just provide an automated system for users to download and run the installer.  

Your use case may not fit into the two scenarios I've described above, but so far they seem to have been quite effective at getting our users the native binaries they need.  We don't try to stop users from using Homebrew in other ways, but as has been stated above, we don't support them.
